Improve SVG Performance

ronkok opened the pull request:
This PR introduces three new classes:  `svgNode`, `pathNode`, and `lineNode`. SVG data is then loaded into the domTree exclusively via instances of these classes.

`innerHTML` is banished.

ronkok commented:
@kevinbarabash I believe that this is one step in a direction where you wan to go. Let me know if I have mis-interpreted things.

ronkok commented:
I've put up a [test page](http://copyasmarkup.com/katex/wideelements.html).

kevinbarabash commented:
This will help big time with https://github.com/Khan/KaTeX/issues/800.  We haven't done much perf work, but perf is one KaTeX's main features so this is very much appreciated.

kevinbarabash commented:
The height on some of the `<svg>` elements on the test page is `1800em` which causes the page to be super long.  This is probably what's causing failures with the screenshot tests.

ronkok commented:
Oops. Multiplied the wrong variable by 1000. I hate it when that happens.

kevinbarabash commented:
@ronkok my assumption is that these changes shouldn't change the output.  I'll run the screenshot tests tomorrow and hopefully that'll shed some light on the failures.

ronkok commented:
When I do this sort of thing, every line of SVG code passes through my hands, and I can't help but tweak things as they go by. In this PR, I changed the `min-length` on many arrows and I rewrote the paths for `\xlinesegment`, `\xequals`, and the left hand side of `\xmapsto`.

I have no idea why `\sqrt` would fail its screenshots. I calculate the same things, but the calculations are done in a different order. Perhaps floating point rounding error?

ronkok commented:
@kevinbarabash Thank you once again for the help with Screenshotter. I know that you're busy just now with flow reviews and I appreciate the time you take to help here.

kevinbarabash commented on src/domTree.js:
```diff
@@ -344,8 +336,164 @@ class symbolNode {
+ * SVG nodes are used to render stretchy wide elements.
+ */
+class svgNode {
+    constructor(children, width, height, x, viewBoxWidth, viewBoxHeight, align) {
```
I think combining all of the arguments after `children` into a single arg called `attributes` would simplify things greatly.  This should simplify `toNode` and `toMarkup` b/c we can iterate through `attributes` instead requiring all the `if` statements.  One slight downside is that now the caller will have to provide `viewBox` as a single value, but I think that's a small price to pay.

ronkok replied:
Okay, I'll do that.

kevinbarabash commented on src/domTree.js:
```diff
@@ -344,8 +336,164 @@ class symbolNode {
+class pathNode {
+    constructor(pathName, alternate) {
+        this.pathName = pathName || "";
+        this.alternate = alternate || "";  // Used only for tall \sqrt
```
The `|| ""` isn't necessary.

kevinbarabash commented on src/domTree.js:
```diff
@@ -344,8 +336,164 @@ class symbolNode {
+}
+
+class lineNode {
+    constructor(x1, y1, x2, y2, strokeWidth) {
```
This could be simplified to (not critical though):
```
constructor(x1 = 0, y1 = 0, x2 = 0, y2 = 0, strokeWidth = 0) {
    Object.assign({x1, y1, x2, y2, strokeWidth});
}
```

ronkok replied:
A single `attributes` arg might be simpler in the constructor. But it also might be more complex in `stretchy.js` I'll look at it.

kevinbarabash replied:
In this case I probably wouldn't bother with a single `attributes` arg even though it deviates from the other constructors.  The reason is that x1, y1, x2, y2 is a well-known ordering of args and you're never going to skip one... it's probably all or none.

kevinbarabash commented on src/stretchy.js:
```diff
@@ -137,309 +92,52 @@ const katexImagesData = {
-<line x1='0' y1='100%' x2='100%' y2='0' stroke-width='0.046em'/>`,
+const katexImagesData = {
+                   //   path(s), minWidth, height, align
+    overrightarrow: [["rightarrow"], 0.888, 522, "xMaxYMin"],
```
Why has the `minWidth` changed for some of these?  Previously it was `0.5` for `overrightarrow`.

ronkok replied:
I discovered `opentype.js`. It made it easier to quickly look at the width of each glyph. So instead of using the width from one glyph and applying it several times, I got a number that matched that specific glyph. For instance, `\overrightarrow` is exactly 888 units width inside the font.

kevinbarabash replied:
Awesome!  Are there situations in which the different `minWidth` will have an impact on the rendering?

ronkok replied:
A function with an empty argument, say `\overrightarrow{}` will be rendered with the min-width. I suppose that min-width might also come into play if the argument is a single, narrow letter like `\imath`.

kevinbarabash commented on src/stretchy.js:
```diff
@@ -453,48 +151,86 @@ const groupLength = function(arg) {
+                // 3 inner SVGs, as in a brace
+                width = ["25.5%", "50%", "25.1%"][i];
+                x = [null, "25%", "74.9%"][i];
+                align = ["xMinYMin", "xMidYMin", "xMaxYMin"][i];
```
Interesting way to grab all of these different values.

kevinbarabash commented on src/svgGeometry.js:
```diff
@@ -0,0 +1,266 @@
+ * It's a storehouse of path geometry for SVG images.
+ */
+
+const path = {
```
I like that these are all paths.

kevinbarabash commented on src/svgGeometry.js:
```diff
@@ -0,0 +1,266 @@
+};
+
+module.exports = {
+    path: path,
```
Are you considering exporting other types of geometry at some point in the future?

ronkok replied:
I can't think of any other types at the moment, but I left it open in case I thought of something later.

kevinbarabash commented on src/svgGeometry.js:
```diff
@@ -0,0 +1,266 @@
+s-117.2-55.8-184.5-66.5c-.7 0-2-.3-4-1-18.7-2.7-76-4.3-172-5H0V214h399571l6 1
+c124.7 8 235 61.7 331 161 31.3 33.3 59.7 72.7 85 118l7 13v35z`,
+
+    rightbraceunder: `M399994 0l6 6v35l-6 11c-56 104-135.3 181.3-238 232-57.3
```
Why do some paths start with `M400000` while others are a little shy of that value?

ronkok replied:
When I wrote the geometry for the absurdly long arrow shafts, I made the end of each shaft take sharp 90° corners. So those points usually landed on x=400000.

In contrast, the KaTeX font glyphs don't have many sharp corners. If the font designer started the path at the right end of the glyph, it was never *exactly* on the edge of the border box. There is always a radius there.

kevinbarabash replied:
I see... so some of these have the non-stretchy part on the right side and then stretchy line extends out to the left?

ronkok replied:
Right. I would start with path geometry from a glyph. Say something like this:
![shortarrow](https://user-images.githubusercontent.com/16403058/30234687-9879c6cc-94b4-11e7-85de-d5ea9433045c.PNG)

Then I would:
1.  Run a little vector transformation script that would shift the whole arrow to the right until its right tip was at x=400000.
2. Edit the points at the left end of the path so they were back at x=0.

The result is a very long arrow:
![longarrow](https://user-images.githubusercontent.com/16403058/30234796-c972b27e-94b5-11e7-95f6-b88407df9b00.PNG)

Okay, so it's hard to see. But on the gray background, that thin line is an arrow and there is an arrow head at one end, I promise. That absurdly long arrow is the inner SVG. The outer SVG is a window that only shows one small section at one end of the long arrow. The arrow head geometry is untouched, relatively. So I'm not really stretching an arrow; I'm really just revealing just enough arrow shaft to match the width of the subject matter.

kevinbarabash commented on src/stretchy.js:
```diff
@@ -453,48 +151,86 @@ const groupLength = function(arg) {
-        depth = imgData[1];
-        imageName = imgData[2];
-        minWidth = imgData[3];
+        const numSvgChildren = imgData[0].length;
```
Instead of indexing into `imgData` for each value, using destructuring would help the reader know which value was which without having to scroll up to the definition of `katexIamgeData`, e.g.
```
const [paths, minWidth, height, align] = katexImagesData[label];
```

ronkok replied:
Will do.

kevinbarabash commented on src/delimiter.js:
```diff
@@ -313,107 +314,59 @@ const makeStackedDelim = function(delim, heightTotal, center, options, mode,
+    let alternate;
+    if (sqrtName === "sqrtTall") {
+        // sqrtTall is from glyph U23B7 in the font KaTeX_Size4-Regular
+        const vertSegment = viewBoxHeight - 54;
```
Where does the `54` come from?

ronkok replied:
From the glyph geometry.  The vertical path length is variable, but it lands on a point that is 54 units from the bottom of the surd.
![landingpoint](https://user-images.githubusercontent.com/16403058/30217127-09c721a8-946a-11e7-89d2-1a55d742a9d0.PNG)

And why, you ask, isn't there another calculation for the vertical length on the left edge? That's because I sequenced the points so that the path starts at the top of that left edge. The final command of a path `z` tells the browser to close the path back to the beginning point. So I've made the browser do that calculation for me.

kevinbarabash replied:
I opened up `KaTeX-Size4` in `opentype.js to look at the coordinates of the surd's outline, but I couldn't find any coordinates near the bottom of the surd where subtracting would give `54`.

<img width="106" alt="screen shot 2017-09-08 at 7 20 29 pm" src="https://user-images.githubusercontent.com/1044413/30234425-fc4f41e8-94ca-11e7-8449-2d829985c30f.png">

ronkok replied:
Correction. That landing point is 14 units from the bottom of the surd (885 - 871 = 14). The path edge in question runs from the bottom of the viniculum to the landing point. The viniculum is 40 units thick. So the math is:

Relative distance from point to point = Overall height - 14 - 40 =  Overall height - 54

kevinbarabash commented on src/buildHTML.js:
```diff
@@ -1598,15 +1598,17 @@ groupTypes.xArrow = function(group, options) {
+        0.5 * arrowBody.height;
+    // 2 mu kern. Ref: amsmath.dtx: #7\if0#2\else\mkern#2mu\fi
+    const upperShift = -options.fontMetrics().axisHeight -
+        0.5 * arrowBody.height - 0.111;
```
Why the change in the shifts?

ronkok replied:
In previous versions, I treated extensible arrows slightly differently than the other SVGs. If an extensible arrow had a total height of 0.522 ems, then the values I returned to `buildHTML` had a `arrowBody.height` of 0.261 and a `arrowBody.depth` of 0.261. Now the code in `stretchy.js` is so array-driven, I found it easier to make things more consistent. Now, even extensible arrows return `.height` with their overall height value and return `.depth` as zero.

So the code inside `buildHTML` had to adjust and split that value in half, so that the arrow is centered on the math axis.

kevinbarabash replied:
> Now the code in stretchy.js is so array-driven, I found it easier to make things more consistent. Now, even extensible arrows return .height with their overall height value and return .depth as zero.

Could you add a comment describing this in stretchy.js?

ronkok replied:
Sure. I'll leave a note in both stretchy.js and in in buildHTML.js.

ronkok commented:
After picking up the review comments, I've refreshed the [test page](http://copyasmarkup.com/katex/wideelements.html)
